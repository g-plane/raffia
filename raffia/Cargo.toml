[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "One parser for CSS/SCSS/Sass/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_macro = { path = "../raffia_macro", version = "0.1" }
serde = { version = "1.0", features = ["derive"], optional = true }
smallvec = { version = "1.9", features = ["const_generics", "serde"] }

[dev-dependencies]
codespan-reporting = "0.11"
criterion = "0.3"
insta = { version = "1.19", features = ["glob", "ron"] }
lightningcss = "1.0.0-alpha.35"
serde_json = "1.0"
swc_common = "0.29.0"
swc_css_ast = "0.114.0"
swc_css_parser = "0.123.0"

[features]
serialize = ["serde"]

[[bench]]
name = "parse_selector"
harness = false

[[bench]]
name = "parse_stylesheet"
harness = false
