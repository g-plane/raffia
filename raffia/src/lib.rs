//! Raffia is a parser can parse CSS, SCSS, Sass (indented syntax) and Less.
//!
//! ## Basic Usage
//!
//! This crate provides a simple API to get started.
//!
//! First, create a parser, give it the source code and specify the syntax,
//! then call the [`parse`](Parser::parse) method:
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css); // syntax can also be `Scss`, `Sass` or `Less`
//! let result = parser.parse::<Stylesheet>();
//! match result {
//!     Ok(ast) => {
//!         // parsed successfully
//!         println!("{:#?}", ast);
//!     }
//!     Err(error) => {
//!         // it failed, error message and position can be accessed via `error`
//!         println!("{:#?}", error);
//!     }
//! }
//! ```
//!
//! ## Advanced Usage
//!
//! ### More Parser Options
//!
//! If you need to control more parser options, you can use [`ParserBuilder`].
//!
//! For example, to collect comments:
//!
//! ```rust
//! use raffia::ParserBuilder;
//!
//! let mut comments = vec![];
//! let builder = ParserBuilder::new("/* comment */ a {}").comments(&mut comments);
//! let mut parser = builder.build();
//! ```
//!
//! By default, syntax is CSS when using parser builder. You can customize it:
//!
//! ```rust
//! use raffia::{ParserBuilder, Syntax};
//!
//! let builder = ParserBuilder::new("a {}").syntax(Syntax::Scss);
//! ```
//!
//! ### Parse Partial Structure
//!
//! Sometimes you don't want to parse a full stylesheet.
//! Say you only need to parse a qualified rule or even a single declaration.
//! All you need to do is to update the generics of the [`parse`](Parser::parse) method.
//!
//! ```rust
//! use raffia::{ast::QualifiedRule, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css);
//! parser.parse::<QualifiedRule>();
//! ```
//!
//! and
//!
//! ```rust
//! use raffia::{ast::Declaration, Parser, Syntax};
//!
//! let mut parser = Parser::new("color: green", Syntax::Css);
//! parser.parse::<Declaration>();
//! ```
//!
//! Not all AST nodes support the usage above;
//! technically, those nodes that implement [`Parse`] trait are supported.
//!
//! ### Retrieve Recoverable Errors
//!
//! There may be some recoverable errors which doesn't affect on producing AST.
//! To retrieve those errors, use [`recoverable_errors`](Parser::recoverable_errors).
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("@keyframes kf { invalid {} }", Syntax::Css);
//! let result = parser.parse::<Stylesheet>();
//! assert!(result.is_ok());
//! println!("{:?}", parser.recoverable_errors());
//! ```
//!
//! ## Serialization
//!
//! Produced AST can be serialized by Serde, but this feature is disabled by default.
//! You need to enable feature `serialize` manually:
//!
//! ```toml
//! raffia = { version = "*", features = ["serialize"] }
//! ```
//!
//! Then you can pass AST to Serde.
//!
//! Note that Raffia only supports serialization. Deserialization isn't supported.

pub use config::{ParserOptions, Syntax};
pub use parser::{Parse, Parser, ParserBuilder};
pub use pos::{Span, Spanned};
pub use span_ignored_eq::SpanIgnoredEq;
pub use tokenizer::token;

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod span_ignored_eq;
mod tokenizer;
mod util;
