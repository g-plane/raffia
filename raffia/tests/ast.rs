use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = match parser.parse::<Stylesheet>() {
            Ok(ast) => {
                let recoverable_errors = parser.recoverable_errors();
                assert!(
                    recoverable_errors.is_empty(),
                    "'{}' has recoverable errors: {recoverable_errors:?}",
                    path.file_name().unwrap().to_str().unwrap(),
                );
                ast
            }
            Err(error) => {
                let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &code);
                let diagnostic = Diagnostic::error()
                    .with_message(error.kind.to_string())
                    .with_labels(vec![Label::primary((), error.span.start..error.span.end)]);
                let mut buffer = Buffer::ansi();
                let config = term::Config::default();
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
                panic!("\n{}", String::from_utf8(buffer.into_inner()).unwrap());
            }
        };

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
