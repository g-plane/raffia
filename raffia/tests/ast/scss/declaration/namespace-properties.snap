---
source: raffia/tests/ast.rs
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    QualifiedRule(
      type: "QualifiedRule",
      selector: SelectorList(
        type: "SelectorList",
        selectors: [
          ComplexSelector(
            type: "ComplexSelector",
            children: [
              CompoundSelector(
                type: "CompoundSelector",
                children: [
                  TagNameSelector(
                    type: "TagNameSelector",
                    name: WqName(
                      type: "WqName",
                      name: Ident(
                        type: "Ident",
                        name: "div",
                        raw: "div",
                        span: Span(
                          start: 0,
                          end: 3,
                        ),
                      ),
                      prefix: None,
                      span: Span(
                        start: 0,
                        end: 3,
                      ),
                    ),
                    span: Span(
                      start: 0,
                      end: 3,
                    ),
                  ),
                ],
                span: Span(
                  start: 0,
                  end: 3,
                ),
              ),
            ],
            span: Span(
              start: 0,
              end: 3,
            ),
          ),
        ],
        span: Span(
          start: 0,
          end: 3,
        ),
      ),
      block: SimpleBlock(
        type: "SimpleBlock",
        statements: [
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: "a",
              raw: "a",
              span: Span(
                start: 8,
                end: 9,
              ),
            ),
            value: [
              SassNestingDeclaration(
                type: "SassNestingDeclaration",
                block: SimpleBlock(
                  type: "SimpleBlock",
                  statements: [
                    Declaration(
                      type: "Declaration",
                      name: Ident(
                        type: "Ident",
                        name: "p1",
                        raw: "p1",
                        span: Span(
                          start: 17,
                          end: 19,
                        ),
                      ),
                      value: [
                        Ident(
                          type: "Ident",
                          name: "q",
                          raw: "q",
                          span: Span(
                            start: 21,
                            end: 22,
                          ),
                        ),
                      ],
                      important: None,
                      lessPropertyMerge: None,
                      span: Span(
                        start: 17,
                        end: 22,
                      ),
                    ),
                    Declaration(
                      type: "Declaration",
                      name: Ident(
                        type: "Ident",
                        name: "b",
                        raw: "b",
                        span: Span(
                          start: 28,
                          end: 29,
                        ),
                      ),
                      value: [
                        SassNestingDeclaration(
                          type: "SassNestingDeclaration",
                          block: SimpleBlock(
                            type: "SimpleBlock",
                            statements: [
                              Declaration(
                                type: "Declaration",
                                name: Ident(
                                  type: "Ident",
                                  name: "p2",
                                  raw: "p2",
                                  span: Span(
                                    start: 39,
                                    end: 41,
                                  ),
                                ),
                                value: [
                                  Ident(
                                    type: "Ident",
                                    name: "q",
                                    raw: "q",
                                    span: Span(
                                      start: 43,
                                      end: 44,
                                    ),
                                  ),
                                ],
                                important: None,
                                lessPropertyMerge: None,
                                span: Span(
                                  start: 39,
                                  end: 44,
                                ),
                              ),
                            ],
                            span: Span(
                              start: 31,
                              end: 51,
                            ),
                          ),
                          span: Span(
                            start: 31,
                            end: 51,
                          ),
                        ),
                      ],
                      important: None,
                      lessPropertyMerge: None,
                      span: Span(
                        start: 28,
                        end: 51,
                      ),
                    ),
                    Declaration(
                      type: "Declaration",
                      name: Ident(
                        type: "Ident",
                        name: "p3",
                        raw: "p3",
                        span: Span(
                          start: 56,
                          end: 58,
                        ),
                      ),
                      value: [
                        Ident(
                          type: "Ident",
                          name: "q",
                          raw: "q",
                          span: Span(
                            start: 60,
                            end: 61,
                          ),
                        ),
                      ],
                      important: None,
                      lessPropertyMerge: None,
                      span: Span(
                        start: 56,
                        end: 61,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 11,
                    end: 66,
                  ),
                ),
                span: Span(
                  start: 11,
                  end: 66,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 8,
              end: 66,
            ),
          ),
        ],
        span: Span(
          start: 4,
          end: 68,
        ),
      ),
      span: Span(
        start: 0,
        end: 68,
      ),
    ),
    QualifiedRule(
      type: "QualifiedRule",
      selector: SelectorList(
        type: "SelectorList",
        selectors: [
          ComplexSelector(
            type: "ComplexSelector",
            children: [
              CompoundSelector(
                type: "CompoundSelector",
                children: [
                  TagNameSelector(
                    type: "TagNameSelector",
                    name: WqName(
                      type: "WqName",
                      name: Ident(
                        type: "Ident",
                        name: "foo",
                        raw: "foo",
                        span: Span(
                          start: 70,
                          end: 73,
                        ),
                      ),
                      prefix: None,
                      span: Span(
                        start: 70,
                        end: 73,
                      ),
                    ),
                    span: Span(
                      start: 70,
                      end: 73,
                    ),
                  ),
                ],
                span: Span(
                  start: 70,
                  end: 73,
                ),
              ),
            ],
            span: Span(
              start: 70,
              end: 73,
            ),
          ),
        ],
        span: Span(
          start: 70,
          end: 73,
        ),
      ),
      block: SimpleBlock(
        type: "SimpleBlock",
        statements: [
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: "bar",
              raw: "bar",
              span: Span(
                start: 78,
                end: 81,
              ),
            ),
            value: [
              SassBinaryExpression(
                type: "SassBinaryExpression",
                left: Ident(
                  type: "Ident",
                  name: "baz",
                  raw: "baz",
                  span: Span(
                    start: 83,
                    end: 86,
                  ),
                ),
                op: SassBinaryOperator(
                  type: "SassBinaryOperator",
                  kind: Plus,
                  span: Span(
                    start: 87,
                    end: 88,
                  ),
                ),
                right: Ident(
                  type: "Ident",
                  name: "bang",
                  raw: "bang",
                  span: Span(
                    start: 89,
                    end: 93,
                  ),
                ),
                span: Span(
                  start: 83,
                  end: 93,
                ),
              ),
              SassNestingDeclaration(
                type: "SassNestingDeclaration",
                block: SimpleBlock(
                  type: "SimpleBlock",
                  statements: [
                    Declaration(
                      type: "Declaration",
                      name: Ident(
                        type: "Ident",
                        name: "bip",
                        raw: "bip",
                        span: Span(
                          start: 100,
                          end: 103,
                        ),
                      ),
                      value: [
                        Ident(
                          type: "Ident",
                          name: "bop",
                          raw: "bop",
                          span: Span(
                            start: 105,
                            end: 108,
                          ),
                        ),
                      ],
                      important: None,
                      lessPropertyMerge: None,
                      span: Span(
                        start: 100,
                        end: 108,
                      ),
                    ),
                    Declaration(
                      type: "Declaration",
                      name: Ident(
                        type: "Ident",
                        name: "bing",
                        raw: "bing",
                        span: Span(
                          start: 114,
                          end: 118,
                        ),
                      ),
                      value: [
                        Ident(
                          type: "Ident",
                          name: "bop",
                          raw: "bop",
                          span: Span(
                            start: 120,
                            end: 123,
                          ),
                        ),
                      ],
                      important: None,
                      lessPropertyMerge: None,
                      span: Span(
                        start: 114,
                        end: 123,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 94,
                    end: 126,
                  ),
                ),
                span: Span(
                  start: 94,
                  end: 126,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 78,
              end: 126,
            ),
          ),
        ],
        span: Span(
          start: 74,
          end: 127,
        ),
      ),
      span: Span(
        start: 70,
        end: 127,
      ),
    ),
    QualifiedRule(
      type: "QualifiedRule",
      selector: SelectorList(
        type: "SelectorList",
        selectors: [
          ComplexSelector(
            type: "ComplexSelector",
            children: [
              CompoundSelector(
                type: "CompoundSelector",
                children: [
                  TagNameSelector(
                    type: "TagNameSelector",
                    name: WqName(
                      type: "WqName",
                      name: Ident(
                        type: "Ident",
                        name: "foo",
                        raw: "foo",
                        span: Span(
                          start: 129,
                          end: 132,
                        ),
                      ),
                      prefix: None,
                      span: Span(
                        start: 129,
                        end: 132,
                      ),
                    ),
                    span: Span(
                      start: 129,
                      end: 132,
                    ),
                  ),
                ],
                span: Span(
                  start: 129,
                  end: 132,
                ),
              ),
            ],
            span: Span(
              start: 129,
              end: 132,
            ),
          ),
        ],
        span: Span(
          start: 129,
          end: 132,
        ),
      ),
      block: SimpleBlock(
        type: "SimpleBlock",
        statements: [
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: "bar",
              raw: "bar",
              span: Span(
                start: 137,
                end: 140,
              ),
            ),
            value: [
              Ident(
                type: "Ident",
                name: "baz",
                raw: "baz",
                span: Span(
                  start: 142,
                  end: 145,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 137,
              end: 145,
            ),
          ),
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: "bang",
              raw: "bang",
              span: Span(
                start: 149,
                end: 153,
              ),
            ),
            value: [
              SassNestingDeclaration(
                type: "SassNestingDeclaration",
                block: SimpleBlock(
                  type: "SimpleBlock",
                  statements: [
                    Declaration(
                      type: "Declaration",
                      name: Ident(
                        type: "Ident",
                        name: "bip",
                        raw: "bip",
                        span: Span(
                          start: 161,
                          end: 164,
                        ),
                      ),
                      value: [
                        Dimension(
                          type: "Dimension",
                          value: Number(
                            type: "Number",
                            value: 1.0,
                            raw: "1",
                            span: Span(
                              start: 166,
                              end: 167,
                            ),
                          ),
                          unit: Ident(
                            type: "Ident",
                            name: "px",
                            raw: "px",
                            span: Span(
                              start: 167,
                              end: 169,
                            ),
                          ),
                          kind: Length,
                          span: Span(
                            start: 166,
                            end: 169,
                          ),
                        ),
                      ],
                      important: None,
                      lessPropertyMerge: None,
                      span: Span(
                        start: 161,
                        end: 169,
                      ),
                    ),
                    Declaration(
                      type: "Declaration",
                      name: Ident(
                        type: "Ident",
                        name: "bop",
                        raw: "bop",
                        span: Span(
                          start: 175,
                          end: 178,
                        ),
                      ),
                      value: [
                        Ident(
                          type: "Ident",
                          name: "bar",
                          raw: "bar",
                          span: Span(
                            start: 180,
                            end: 183,
                          ),
                        ),
                      ],
                      important: None,
                      lessPropertyMerge: None,
                      span: Span(
                        start: 175,
                        end: 183,
                      ),
                    ),
                    Declaration(
                      type: "Declaration",
                      name: Ident(
                        type: "Ident",
                        name: "blat",
                        raw: "blat",
                        span: Span(
                          start: 189,
                          end: 193,
                        ),
                      ),
                      value: [
                        SassNestingDeclaration(
                          type: "SassNestingDeclaration",
                          block: SimpleBlock(
                            type: "SimpleBlock",
                            statements: [
                              Declaration(
                                type: "Declaration",
                                name: Ident(
                                  type: "Ident",
                                  name: "baf",
                                  raw: "baf",
                                  span: Span(
                                    start: 195,
                                    end: 198,
                                  ),
                                ),
                                value: [
                                  Ident(
                                    type: "Ident",
                                    name: "bort",
                                    raw: "bort",
                                    span: Span(
                                      start: 199,
                                      end: 203,
                                    ),
                                  ),
                                ],
                                important: None,
                                lessPropertyMerge: None,
                                span: Span(
                                  start: 195,
                                  end: 203,
                                ),
                              ),
                            ],
                            span: Span(
                              start: 194,
                              end: 204,
                            ),
                          ),
                          span: Span(
                            start: 194,
                            end: 204,
                          ),
                        ),
                      ],
                      important: None,
                      lessPropertyMerge: None,
                      span: Span(
                        start: 189,
                        end: 204,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 155,
                    end: 205,
                  ),
                ),
                span: Span(
                  start: 155,
                  end: 205,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 149,
              end: 205,
            ),
          ),
        ],
        span: Span(
          start: 133,
          end: 206,
        ),
      ),
      span: Span(
        start: 129,
        end: 206,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 207,
  ),
)
